openmc/model/model.py

```diff
@@ -289,23 +514,395 @@ def run(self, **kwargs):
+            MPI execute command and any additional MPI arguments to pass,
+            e.g. ['mpiexec', '-n', '8'].
+            This only applies to the case when not using the C-API.
+        cwd : str, optional
+            Path to working directory to run in. Defaults to the current
+            working directory.
+        apply_volumes : bool, optional
+            Whether apply the volume calculation results from this calculation
+            to the model. Defaults to applying the volumes.
+        """
+
+        if len(self.settings.volume_calculation) == 0:
+            # Then there is no volume calculation specified
+            raise ValueError("The Settings.volume_calculation attribute must"
+                             " be specified before executing this method!")
+
+        with openmc.change_directory(Path(cwd)):
+            if self.C_init:
+                if threads is not None:
+                    msg = "Threads must be set via Model.c_init(...)"
+                    raise ValueError(msg)
+                if mpi_args is not None:
+                    msg = "The MPI environment must be set otherwise such as" \
+                        "with the call to mpi4py"
+                    raise ValueError(msg)
+
+                # Apply the output settings
+                if not output:
+                    init_verbosity = openmc.lib.settings.verbosity
+                    if not output:
+                        openmc.lib.settings.verbosity = 1
+
+                # Compute the volumes
+                openmc.lib.calculate_volumes()
+
+                # Reset the output verbosity
+                if not output:
+                    openmc.lib.settings.verbosity = init_verbosity
+            else:
+                self.export_to_xml()
+                openmc.calculate_volumes(threads=threads, output=output,
+                                         openmc_exec=openmc_exec,
+                                         mpi_args=mpi_args)
+
+            # Now we apply the volumes
+            if apply_volumes:
+                # Load the results
+                f_names = \
+                    ["volume_{}.h5".format(i + 1)
+                     for i in range(len(self.settings.volume_calculations))]
+                vol_calcs = [openmc.VolumeCalculation.load_results(f_name)
+                             for f_name in f_names]
+                # And now we can add them to the model
+                for vol_calc in vol_calcs:
+                    # First add them to the Python side
+                    self.geometry.add_volume_information(vol_calc)
+
+                    # And now repeat for the C-API
+                    if vol_calc.domain == 'material':
+                        # Then we can do this in the C-API
+                        for domain_id in vol_calc.domains:
+                            self.update_material_volumes(
+                                [domain_id], vol_calc.volumes[domain_id])
```

Comment:
We should probably encapsulate the ability to update volumes in `openmc.lib.calculate_volumes` itself. What you have here works for now though